lib/figaro_yaml_validator.rb

```diff
@@ -0,0 +1,30 @@
+class FigaroYamlValidator
+  FILE_PATH = File.join(Rails.root, 'config', 'application.yml')
+  FIGARO_YAML = YAML.load(IO.read(FILE_PATH))
+
+  def validate(yaml_data = FIGARO_YAML)
+    bad_keys = []
+    check_for_bad_values(yaml_data, bad_keys)
+    return unless bad_keys.any?
+    raise warning(bad_keys).tr("\\\n", ' ')
+  end
+
+  private
+
+  def check_for_bad_values(yaml_data, bad_keys)
+    yaml_data.map do |key, value|
+      if value.is_a?(Hash)
+        check_for_bad_values(value, bad_keys)
+      elsif value == 'yes' || value == 'no'
+        bad_keys << key
+      end
+    end
+  end
+
+  def warning(bad_keys)
+    <<~HEREDOC
+      You have invalid values for #{bad_keys.to_sentence} in
+      config/application.yml. Please change them to true or false
+    HEREDOC
+  end
+end
```

Comment:
Was there something wrong with my solution? This seems more complicated than it needs to be. Also, this looks at all the values in all the Rails.env sections, which can cause duplicates. For example, if both the development and production entries for a key have an incorrect value, the message will look like this:

```
You have invalid values for allow_third_party_auth and allow_third_party_auth in config/application.yml. Please change them to true or false
```
Similarly, if the production value is incorrect, but the development value is correct, raising an error for the production entry when running the app in development might cause confusion, which is why I added `.fetch(Rails.env)` in my solution so that the validation only runs against the current environment.

By using `.load_file`, you get a Hash right away and you don't have to do any type checks.

The one thing I missed in my solution was an `if result` for the raise so that it only raises if the validation fails.